### PR TITLE
Tags refactor

### LINKED ISSUE #2548: [FR] Tags update (combine sample tags and label tags into a single group)
### Proposal Summary

We should update the SAMPLE TAGS and LABEL TAGS sections in the App to instead show a single TAGS section that contains tags and label tags sections that use the existing ListField(StringField) filtering widgets to encapsulate the existing sample and label tags filtering UIs, respectively.

The benefits of this are numerous:

- Removes custom code that currently exists for sample/label tags sections
- Enables us to leverage the filter type UI in to provide more options like showing samples that do not contain a given tag

Notes

This will require migrating existing app_config.sidebar_groups to use the new tags group name

### PR DESCRIPTION
**Pull request by lanzhenw**
Resolves #2548 

https://user-images.githubusercontent.com/17770824/220705304-cdfbef11-fef3-43c6-97f2-a7b8b6f12e51.mp4

## What changes are proposed in this pull request?

We update the `Sample Tags` and `Label Tags` sections in the App into one combined group `Tags`, which contains label tags and sample tags sections as ListField(StringField) filters. 

- The update removes custom UI code for sample/label tags in recoil.
- The update allows users to match/omit samples by sample tags; 
- The update allows users to select/exclude labels and match/omit samples by label tags;

## How is this patch tested? If it is not, please explain why.

Locally with quickstart data

### Is this a user-facing change that should be mentioned in the release notes?

-   [ ] No. You can skip the rest of this section.
-   [x] Yes. Give a description of this change to be included in the release
        notes for FiftyOne users.

(Details in 1-2 sentences. You can just refer to another PR with a description
if this PR is part of a larger change.)

### What areas of FiftyOne does this PR affect?

-   [x] App: FiftyOne application changes
-   [ ] Build: Build and test infrastructure changes
-   [ ] Core: Core `fiftyone` Python library changes
-   [ ] Documentation: FiftyOne documentation changes
-   [ ] Other

### CODE REVIEW

**Comment by lanzhenw** [comment]
Status update: Screenshot 2023-01-23 at 11 32 43 AM

**Review by benjaminpkane — CHANGES_REQUESTED** [review]
Code looks really good!

One thing I noticed in testing is the sidebar is suspending while tag data is loading which is a regression. We want to show the sidebar immediately, which means any recoil value that may suspend (is async with a network request) must be wrapped by a `Suspense` as closely as possible to the component showing that data. Or you can use `useRecoilValueLoadable`

File: app/packages/core/src/components/SamplesContainer.tsx [hunk 1]
```diff
@@ -31,71 +31,42 @@ function SamplesContainer() {
...
-            disabled: isTag || isLabelTag || disabled.has(entry.path),
+            children: (
+              <Entries.FilterablePath
+                entryKey={key}
+                disabled={isDisabled}
+                group={group}
+                key={key}
+                modal={false}
+                path={entry.path}
+                childPaths={entry.childPaths}
```
[the remark this answers is not in the document]
Comment (lanzhenw) [reply app/packages/core/src/components/SamplesContainer.tsx]: cleaned it

File: app/packages/state/src/recoil/sidebar.ts [hunk 2]
```diff
@@ -65,6 +65,7 @@ export interface PathEntry {
   kind: EntryKind.PATH;
   path: string;
   shown: boolean;
+  childPaths?: string[];
```
[the remark this answers is not in the document]
Comment (lanzhenw) [reply app/packages/state/src/recoil/sidebar.ts]: I just cleaned it. Thanks for catching it!

**Review by brimoor — CHANGES_REQUESTED** [review]
@lanzhenw looking good! Here's a couple things I saw:

- Can you update the implementation so that sample tag counts are not computed by default when sidebar mode is "fast"? See below. The intention is that counts are only computed when the sidebar mode is "all" OR the filter tray for that field is expanded
- Can you update the default sidebar groups so that sample tags are above label tags?

Screen Shot 2023-03-07 at 11 20 21 AM

**Comment by lanzhenw** [comment]
> @lanzhenw what's the intended behavior when trying to reorder `tags` and `label tags` in the sidebar? Previously this was not possible (individual tags could not be dragged). Now I am able to, for example, drag `tags` into the `LABELS` section, but then I cannot drag it back.
> 
> Shall we make it so that tags paths cannot be reorganized?

Yes, since tags paths are specially coded, I just made label tag and sample tag undraggable.

**Comment by brimoor** [comment]
@lanzhenw I'm seeing a couple minor things and one key issue:

## Key issue

- As shown in the screenshot below, sometimes when I select a label tag to filter by, the sidebar and samples count update as expected, but the grid doesn't render. If I click the bookmark icon, then the grid renders correctly. Potentially related? This is a dataset with multiple label fields (`Classification`, `Detections`, and `Polylines`)

Screen Shot 2023-03-07 at 11 23 05 PM

## Minor issues

- The tags fields cannot actually be dragged, but when you hover over them the "drag mode" still activates, suggesting that you should be able to drag. Can you disable this too?
- When there are no label tags, you cannot click the `label tags` visibility checkbox, whereas other fields with a count of zero still allow you to click the checkbox

Screen Shot 2023-03-07 at 11 22 22 PM

## Not strictly related to this PR, but can you look at it? 🤗 

- Certain fields (`Polylines` and `StringField`) with a count of zero allow their filter tray to be expanded, showing a "No results" state like shown below. However, other fields (`Detections` and `IntField`) immediately collapse after you try to open them if the count is zero. This makes it a bit confusing whether the field is actually expanded or not. It would be preferable if all filter trays could be opened, showing an appropriate "No results" state if necessary

Screen Shot 2023-03-07 at 11 24 02 PM

**Comment by lanzhenw** [comment]
@brimoor Thanks for catching the issues. Got them fixed.
Key issue: 
https://user-images.githubusercontent.com/17770824/223880609-e788426f-6484-48d5-9b9d-69e9c1a50bcf.mov
Minor issues:
https://user-images.githubusercontent.com/17770824/223880802-ba949423-77b6-4387-b1b6-c7ccf4ba7b11.mov

**Review by brimoor — APPROVED** [review]
I defer to @benjaminpkane for final code approval, but usability is great. Nice work @lanzhenw! 💪

File: app/packages/core/src/components/Filters/categoricalFilter/CategoricalFilter.tsx [hunk 3]
```diff
@@ -70,17 +72,24 @@ const categoricalSearchResults = selectorFamily<
       }
 
       const noneCount = get(fos.noneCount({ path, modal, extended: false }));
+      const isLabelTag = path.startsWith("_label_tags");
+      const labels = get(labelTagsCount({ modal, extended: false }));
```
Comment (benjaminpkane) [inline app/packages/core/src/components/Filters/categoricalFilter/CategoricalFilter.tsx:76]: It's better to subscribe to `labelTagsCount` only when `isLabelTag` is true

File: app/packages/core/src/components/Sidebar/Entries/EntryCounts.tsx [hunk 4]
```diff
@@ -72,37 +71,35 @@ const labelTagCount = selectorFamily<
...
-        toggle
-          ? new Set([tag, ...current])
-          : new Set([...current].filter((t) => t !== tag))
+      const r1 = get(
+        fos.cumulativeCounts({
+          ...fos.MATCH_LABEL_TAGS,
+          ...props,
+        })
       );
+      if (!r1) return { count: 0, results: [] };
```
Comment (benjaminpkane) [inline app/packages/core/src/components/Sidebar/Entries/EntryCounts.tsx:99]: Maybe these variable names could be more descriptive? Just a suggestion

**Review by benjaminpkane — APPROVED** [review]
LGTM! I don't believe my updates broke anything, but please let me know if I did.

**Pull request merged**